sdk/storage/azure-storage-common/inc/azure/storage/common/paged_response.hpp

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <cstdlib>
+#include <string>
+
+#include <azure/core/context.hpp>
+#include <azure/core/http/raw_response.hpp>
+
+namespace Azure {
+
+template <class Derived> class PagedResponse {
+public:
+  std::string NextPageToken;
+  std::string CurrentPageToken;
+  std::unique_ptr<Azure::Core::Http::RawResponse> RawResponse;
+
+  bool HasMorePages() const { return m_hasMorePages; }
+
+  void MoveToNextPage(const Azure::Core::Context& context = Azure::Core::Context())
+  {
+    static_assert(
+        std::is_base_of<PagedResponse, Derived>::value,
+        "The template argument \"Derived\" should derive from PagedResponse<Derived>.");
+
+    if (NextPageToken.empty())
+    {
+      m_hasMorePages = false;
+      return;
+    }
+    CurrentPageToken = NextPageToken;
+    static_cast<Derived*>(this)->OnNextPage(context);
+  }
+
+protected:
+  PagedResponse() = default;
+  PagedResponse(PagedResponse&&) = default;
+  PagedResponse& operator=(PagedResponse&&) = default;
```

Comment:
@CaseyCarter why do we define these, default ctor/move ctor/move assignment functions?

And also, why did we decide to make them protected?

Reply:
Because we want to make base class `PagedResponse` and its derived class movable and move assignable, but not copiable.

Reply:
This is useful in the implementation of `OnNextPage`.

Reply:
> and its derived class movable and move assignable, but not copiable.

Why don't we want it to be copiable? Just trying to understand what motivates such decisions.

Reply:
unique pointer is not copiable. Because otherwise, it wouldn't be **unique**. So any struct with unique pointer as member is not copiable.

Reply:
Ah, yes, thanks!

Reply:
And why make them protected rather than public or private?

Reply:
Because `PagedResponse` is never supposed to be used directly. We only use its derived class. So by making it protected, we can instantiate a derived class and initialize the base class in the constructor of derived class.